Extract `models` into a separate e2e CI group

nemanjaglumac opened the pull request:
Status:
This is a blocker for the copy changes and should be merged asap.

Extracts models into a separate e2e group and splits the large spec into a few smaller logical units.
Reduces total running time by 2-3 minutes for a `question` e2e group, which is the longest running test in our CI.

nemanjaglumac committed:
Extract datasets into a separate e2e group

nemanjaglumac committed:
Split datasets tests into smaller logical units

kulyk commented on frontend/test/metabase/scenarios/datasets/helpers/e2e-datasets-helpers.js:
```diff
@@ -0,0 +1,128 @@
+    .click();
+}
+
+export function selectDimensionOptionFromSidebar(name) {
+  cy.get("[data-testid=dimension-list-item]")
+    .contains(name)
+    .click();
+}
+
+export function openDetailsSidebar() {
+  cy.findByTestId("saved-question-header-button").click();
+}
+
+export function getDetailsSidebarActions() {
+  return cy.findByTestId("question-action-buttons");
+}
```
These functions can be used for questions too. Would it make sense to move them to something like `helpers/e2e-question-helpers`?

kulyk commented on frontend/test/metabase/scenarios/datasets/helpers/e2e-datasets-helpers.js:
```diff
@@ -0,0 +1,128 @@
+  });
+}
+
+export function getCollectionItemRow(itemName) {
+  return cy.findByText(itemName).closest("tr");
+}
```
Should we move that to `helpers/e2e-collection-helpers`?

nemanjaglumac replied:
Now that I think about this, it can fail if there are two collection items with the same name (which is quite possible).
In general, until we start using helpers in a lot of different specs and across different areas I am always reluctant to extract them into "global" helpers.

kulyk commented on frontend/test/metabase/scenarios/datasets/helpers/e2e-datasets-helpers.js:
```diff
@@ -0,0 +1,128 @@
+  return cy.findByText(itemName).closest("tr");
+}
+
+export function selectFromDropdown(option, clickOpts) {
+  popover()
+    .findByText(option)
+    .click(clickOpts);
+}
```
Should we move that to `helpers/e2e-ui-elements-helpers`?

nemanjaglumac replied:
The same concern (like [this one](https://github.com/metabase/metabase/pull/19650#discussion_r784247566)) goes here. The very common example is "Created At" which can appear 2 times in a single dropdown. These helpers need to be more robust if we want to use them globally.

nemanjaglumac replied:
Btw, this same helper has been defined in at least 3-4 different places, not related to these changes.

kulyk commented on frontend/test/metabase/scenarios/datasets/helpers/e2e-datasets-helpers.js:
```diff
@@ -0,0 +1,128 @@
+    .click(clickOpts);
+}
+
+export function joinTable(table) {
+  cy.icon("join_left_outer").click();
+  selectFromDropdown(table);
+}
```
Should we move that to `helpers/e2e-notebook-helpers`?

kulyk commented on frontend/test/metabase/scenarios/datasets/helpers/e2e-datasets-helpers.js:
```diff
@@ -0,0 +1,128 @@
+  selectFromDropdown(table);
+}
+
+export function testDataPickerSearch({
+  inputPlaceholderText,
+  query,
+  datasets = false,
+  cards = false,
+  tables = false,
+} = {}) {
+  cy.findByPlaceholderText(inputPlaceholderText).type(query);
+  cy.wait("@search");
+
+  cy.findAllByText(/Dataset in/i).should(datasets ? "exist" : "not.exist");
+  cy.findAllByText(/Saved question in/i).should(cards ? "exist" : "not.exist");
+  cy.findAllByText(/Table in/i).should(tables ? "exist" : "not.exist");
+
+  cy.icon("close").click();
+}
```
I'm not sure any other that would actually need that helper. WDYT about keeping it in the test file?

nemanjaglumac replied:
Tbh, at this point I wasn't too concerned about where would things live, because we need to add more dataset tests and the cleanup phase is inevitable.

Do you think that some repro in the future might use this helper?

kulyk commented on frontend/test/metabase/scenarios/datasets/datasets.cy.spec.js:
```diff
@@ -354,13 +368,13 @@ describe("scenarios > datasets", () => {
           .type("D1");
         cy.findByLabelText("Description")
           .clear()
-          .type("Some helpful dataset description");
+          .type("foo");
         cy.button("Save").click();
```
Just out of curiosity: does this change saves CI time? If so, I'd prefer shorter input strings in the future too

nemanjaglumac replied:
It saves around 3s from this single test alone.

kulyk replied:
Wow! Good to know, thank you!

kulyk reviewed (commented):
That's awesome, thank you! Just have a few ideas on how to structure the helpers

kulyk commented:
@nemanjaglumac do we need the "backport" label here as datasets are not released yet?

nemanjaglumac commented:
> @nemanjaglumac do we need the "backport" label here as datasets are not released yet?

Good point.
We don't :)

nemanjaglumac committed:
Keep `getCollectionItemRow` helper scoped to a single file

nemanjaglumac committed:
Keep `testDataPickerSearch` helper scoped to a single file

nemanjaglumac committed:
Update the name from `datasets` to `models`

nemanjaglumac commented:
NOTE:
It has been decided that `datasets` are going to be officially called `models`. I've updated test group and the file names.
However, a lot of the helper functions are still using the term `datasets`. Until we consolidate everything, it's going to be a bit messy. 

@kulyk I've updated some of the helpers, according to your review. Other are used in way too many places and some even have 2-3-4 different implementations across all test files. In order to NOT pollute this PR, I'll take care of the helper files cleanup separately. Hope that makes sense.

gusaiani reviewed (approved):
Looks good, onwards 👍!

kulyk commented:
> NOTE:
> 
> It has been decided that `datasets` are going to be officially called `models`. I've updated test group and the file names.
> 
> However, a lot of the helper functions are still using the term `datasets`. Until we consolidate everything, it's going to be a bit messy. 
> 
> 
> 
> @kulyk I've updated some of the helpers, according to your review. Others are used in way too many places and some even have 2-3-4 different implementations across all test files. In order to NOT pollute this PR, I'll take care of the helper files cleanup separately. Hope that makes sense.

That makes perfect sense, thank you 🙏
